### PR TITLE
 Show a warning when registering for several courses at the same time

### LINKED ISSUE #864: Предупреждение при регистрации на несколько курсов одновременно
Если студент в активном статусе, на активном курсе, пытается зарегистрироваться ещё на один курс, сделать модальное с предупрждением, что проходить несколько курсов одновременно не рекомендуется.

### LINKED ISSUE #909: Неправильно отображается колличество студентов в профайле ментора
photo5363909004876690582

Если ментор только зарегистрировался на курс и у него нету студентов, в профайле отображается что у ментора имеется 1 студент.

### LINKED ISSUE #910: Регистрация на закрытый курс
Если перейти по прямой ссылке вида `https://app.rs.school/registry/student?course=angular-2021Q3` для регистрации на конкретный курс, то можно успешно зарегистрироваться на любой курс, даже завершенный.

### PR DESCRIPTION
**Pull request by Pulya10c**
fixes #864
fixes #909
fixes #910

### CODE REVIEW

File: client/src/modules/Registry/pages/Student/index.tsx [hunk 1]
```diff
@@ -26,24 +27,40 @@ export type Props = {
   session: Session;
 };
 
+interface IdName {
```
Comment (apalchys) [inline client/src/modules/Registry/pages/Student/index.tsx:30]: please create a type in such case

File: client/src/modules/Registry/pages/Student/index.tsx [hunk 2]
```diff
@@ -26,24 +27,40 @@ export type Props = {
...
   const [initialData, setInitialData] = useState({} as Partial<UserFull>);
 
   useAsync(async () => {
     setLoading(true);
     const userService = new UserService();
     const courseService = new CoursesService();
-    const [profile, courses] = await Promise.all([userService.getMyProfile(), courseService.getCourses()]);
+    const [profile, profileInfo, courses] = await Promise.all([
+      userService.getMyProfile(),
+      userService.getProfileInfo(githubId),
```
Comment (apalchys) [inline client/src/modules/Registry/pages/Student/index.tsx:52]: why do you need to pull profile info data? Everything should be available in `props.session`. Isn't it?

File: client/src/modules/Registry/pages/Student/index.tsx [hunk 3]
```diff
@@ -259,18 +310,32 @@ function sortByStartDate(a: Course, b: Course) {
   return a.startDate.localeCompare(b.startDate);
 }
 
-function isCourseOpenForRegistry(course: Course) {
+function enrolledOtherCourses(studentStats: StudentStats[] | undefined, courses: Course[]) {
+  return (
+    studentStats?.reduce((acc, { isExpelled, isCourseCompleted, courseId }) => {
+      const course = courses?.find(el => el.id === courseId);
+      if (!isExpelled && !isCourseCompleted && course && !course?.completed) {
```
Comment (apalchys) [inline client/src/modules/Registry/pages/Student/index.tsx:317]: `course && !course?.completed` 

you can replace it with `course?.completed === false`

File: client/src/modules/Registry/pages/Student/index.tsx [hunk 4]
```diff
@@ -259,18 +310,32 @@ function sortByStartDate(a: Course, b: Course) {
...
+      return acc;
+    }, [] as IdName[]) ?? []
+  );
+}
+
+function isCourseOpenForRegistry(registeredCourses: IdName[]) {
   // invite only courses do not open for public registration
-  if (course.inviteOnly || course.completed) {
+  return (course: Course) => {
+    if (course.inviteOnly || course.completed || registeredCourses.find(({ id }) => id === course.id)) {
```
Comment (apalchys) [inline client/src/modules/Registry/pages/Student/index.tsx:328]:
```suggestion
    if (course.inviteOnly || course.completed || registeredCourses.some(({ id }) => id === course.id)) {
```

File: client/src/modules/Registry/pages/Student/index.tsx [hunk 5]
```diff
@@ -26,24 +27,40 @@ export type Props = {
...
   const [form] = Form.useForm();
   const update = useUpdate();
   const [loading, setLoading] = useState(false);
   const [submitted, setSubmitted] = useState(false);
   const [courses, setCourses] = useState([] as Course[]);
+  const [registeredForCourses, setRegisteredForCourses] = useState([] as IdName[]);
   const [location, setLocation] = useState(null as Location | null);
   const [initialData, setInitialData] = useState({} as Partial<UserFull>);
 
   useAsync(async () => {
```
Comment (apalchys) [inline client/src/modules/Registry/pages/Student/index.tsx:46]: Also maybe please consider moving it to a separate hook. It looks too complicated now for a component

File: client/src/modules/Registry/pages/Student/index.tsx (hunk 2 above)
Comment (Pulya10c) [reply client/src/modules/Registry/pages/Student/index.tsx:52]: Can we get the values for the course isExpelled and isCourseCompleted from session? I have not found a way to do this.

File: client/src/modules/Registry/pages/Student/index.tsx (hunk 5 above)
Comment (Pulya10c) [reply client/src/modules/Registry/pages/Student/index.tsx]: done

File: client/src/modules/Registry/pages/Student/index.tsx (hunk 4 above)
Comment (Pulya10c) [reply client/src/modules/Registry/pages/Student/index.tsx]: done

File: client/src/modules/Registry/pages/Student/index.tsx (hunk 3 above)
Comment (Pulya10c) [reply client/src/modules/Registry/pages/Student/index.tsx]: done

File: client/src/modules/Registry/pages/Student/index.tsx (hunk 1 above)
Comment (Pulya10c) [reply client/src/modules/Registry/pages/Student/index.tsx]: done

**Review by apalchys — APPROVED** [review]

**Pull request merged**